sx128x.cpp

```diff
@@ -0,0 +1,882 @@
+
+extern SPIClass spiModem;
+#define SPI spiModem
+
+extern SPIClass SPI;
+
+#define MAX_PKT_LENGTH           255
+
+sx128x::sx128x() :
+  _spiSettings(8E6, MSBFIRST, SPI_MODE0),
+  _ss(LORA_DEFAULT_SS_PIN), _reset(LORA_DEFAULT_RESET_PIN), _dio0(LORA_DEFAULT_DIO0_PIN), _rxen(LORA_DEFAULT_RXEN_PIN), _busy(LORA_DEFAULT_BUSY_PIN),
+  _frequency(0),
+  _txp(0),
+  _sf(0x50),
+  _bw(0x34),
+  _cr(0x01),
+  _packetIndex(0),
+  _preambleLength(18),
+  _implicitHeaderMode(0),
+  _payloadLength(255),
+  _crcMode(0),
+  _fifo_tx_addr_ptr(0),
+  _fifo_rx_addr_ptr(0),
+  _packet({0}),
+  _rxPacketLength(0),
+  _preinit_done(false),
+  _onReceive(NULL)
+{
+  // overide Stream timeout value
+  setTimeout(0);
+}
+
+bool sx128x::preInit() {
+  // setup pins
+  pinMode(_ss, OUTPUT);
+  // set SS high
+  digitalWrite(_ss, HIGH);
+  
+  SPI.begin();
+
+  // check version (retry for up to 2 seconds)
+  long start = millis();
+
+  uint8_t version_msb;
+  uint8_t version_lsb;
+
+  while (((millis() - start) < 2000) && (millis() >= start)) {
+
+      version_msb = readRegister(REG_FIRM_VER_MSB);
+      version_lsb = readRegister(REG_FIRM_VER_LSB);
+
+      if ((version_msb == 0xB7 && version_lsb == 0xA9) || (version_msb == 0xB5 && version_lsb == 0xA9)) {
+          break;
+      }
+      delay(100);
+  }
+  if ((version_msb != 0xB7 || version_lsb != 0xA9) && (version_msb != 0xB5 || version_lsb != 0xA9)) {
+      return false;
+  }
+
+  _preinit_done = true;
+  return true;
+}
+
+uint8_t ISR_VECT sx128x::readRegister(uint16_t address)
+{
+  return singleTransfer(OP_READ_REGISTER_8X, address, 0x00);
+}
+
+void sx128x::writeRegister(uint16_t address, uint8_t value)
+{
+    singleTransfer(OP_WRITE_REGISTER_8X, address, value);
+}
+
+uint8_t ISR_VECT sx128x::singleTransfer(uint8_t opcode, uint16_t address, uint8_t value)
+{
+    waitOnBusy();
+
+    uint8_t response;
+
+    digitalWrite(_ss, LOW);
+
+    SPI.beginTransaction(_spiSettings);
+    SPI.transfer(opcode);
+    SPI.transfer((address & 0xFF00) >> 8);
+    SPI.transfer(address & 0x00FF);
+    if (opcode == OP_READ_REGISTER_8X) {
+        SPI.transfer(0x00);
+    }
+    response = SPI.transfer(value);
+    SPI.endTransaction();
+
+    digitalWrite(_ss, HIGH);
+
+    return response;
+}
+
+void sx128x::rxAntEnable()
+{
+    if (_txen != -1) {
+        digitalWrite(_txen, LOW);
+    }
+    if (_rxen != -1) {
+        digitalWrite(_rxen, HIGH);
+    }
+}
+
+void sx128x::txAntEnable()
+{
+    if (_txen != -1) {
+        digitalWrite(_txen, HIGH);
+    }
+    if (_rxen != -1) {
+        digitalWrite(_rxen, LOW);
+    }
+}
+
+void sx128x::loraMode() {
+    // enable lora mode on the SX1262 chip
+    uint8_t mode = MODE_LONG_RANGE_MODE_8X;
+    executeOpcode(OP_PACKET_TYPE_8X, &mode, 1);
+}
+
+void sx128x::waitOnBusy() {
+    unsigned long time = millis();
+    if (_busy != -1) {
+        while (digitalRead(_busy) == HIGH)
+        {
+            if (millis() >= (time + 300)) {
+                pinMode(LED_BLUE, OUTPUT);
```

Comment:
Is  this LED control part of the indications for users, or a debug function?

Reply:
Sorry buddy, I was worried something like that had slipped in there! Yes, this was for my testing when debugging, I usually put comments around stuff like this so I remember to remove it, but it seems I forgot in this case. Reverted by a179304.

Reply:
No worries! Was just wondering whether it was actually the point or not ;)